Batch Control Plane 2021-06-01 update

paterasMSFT opened the pull request:
[Swagger change PR](https://github.com/Azure/azure-rest-api-specs/pull/14862)

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

paterasMSFT committed:
Generated control plane

Added generated descriptions.

paterasMSFT committed:
Updated version

PR SDK updates

Updated version

sdkinfo

paterasMSFT committed:
Session Records

Updated test recordings

Updated recordings

Updated recordings

Updated session records

paterasMSFT committed:
Changelog lint fixes.

Added markdown config

Ignore MD013

Change log update

Expanded change log.

paterasMSFT committed:
Tests

AutoStorage AuthenticationMode memoery test

Switched to static resource name.

Scenario tests

Added ListOutboundNetworkDependenciesEndpoints test

OsDisk scenario test

Tests AllowedAuthenticationModes

ComputeNodeIdentityReference test

paterasMSFT committed:
Cleanup

allenjzhang commented:
/azp run net - mgmt - ci

paterasMSFT commented:
@allenjzhang CI passing now

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Generated/Models/BatchAccountIdentity.cs:
```diff
@@ -43,10 +44,8 @@ public BatchAccountIdentity()
-        /// '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}'.</param>
-        public BatchAccountIdentity(ResourceIdentityType type, string principalId = default(string), string tenantId = default(string), IDictionary<string, BatchAccountIdentityUserAssignedIdentitiesValue> userAssignedIdentities = default(IDictionary<string, BatchAccountIdentityUserAssignedIdentitiesValue>))
+        /// associated with the Batch account.</param>
+        public BatchAccountIdentity(ResourceIdentityType type, string principalId = default(string), string tenantId = default(string), IDictionary<string, UserAssignedIdentities> userAssignedIdentities = default(IDictionary<string, UserAssignedIdentities>))
```
breaking sdk change

markcowl replied:
still a breaking change - the constructor has removed parameters.

markcowl replied:
I missed the added class, was looking in customizations, but this works.

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Generated/Models/BatchAccountIdentityUserAssignedIdentitiesValue.cs:
```diff
@@ -1,60 +0,0 @@
-// <auto-generated>
```
same breaking sdk change as above

markcowl replied:
It looks like these two types are functionally equivalent.  Can we simply add this type back in as extending the new 'UserAssignedIdentities' type to avoid the breaking change?

markcowl replied:
Still a breaking change, this class is still removed with no mitigation

paterasMSFT replied:
Added manually here: sdk/batch/Microsoft.Azure.Management.Batch/src/Models/BatchAccountIdentityUserAssignedIdentitiesValue.cs

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Generated/Models/BatchPoolIdentity.cs:
```diff
@@ -36,10 +36,8 @@ public BatchPoolIdentity()
-        /// associated with the Batch pool. The user identity dictionary key
-        /// references will be ARM resource ids in the form:
-        /// '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}'.</param>
-        public BatchPoolIdentity(PoolIdentityType type, IDictionary<string, BatchPoolIdentityUserAssignedIdentitiesValue> userAssignedIdentities = default(IDictionary<string, BatchPoolIdentityUserAssignedIdentitiesValue>))
```
same comment - replacing the identity type is a breaking change, but since these types appear to be equivalent, we may be able to fix by re-introducing the old type.

markcowl replied:
still a breaking change - was there somne mitigation that you intended?

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Generated/Models/ContainerRegistry.cs:
```diff
@@ -30,16 +29,20 @@ public ContainerRegistry()
+        /// <param name="identityReference">The reference to the user assigned
+        /// identity to use to access an Azure Container Registry instead of
+        /// username and password.</param>
+        public ContainerRegistry(string registryServer = default(string), string userName = default(string), string password = default(string), ComputeNodeIdentityReference identityReference = default(ComputeNodeIdentityReference))
```
hmmm, since these parameters are all strings, this looks like a breaking change we cannot mitigate.

markcowl replied:
I think the easiest way to deal with this may be to add the x-ms-external directive to this class in the C# autorest readme - this would allow you to provide this class is customizations, with a non-breaking parameter ordering.  If we don't want to make this change, or something similar, we will have to request a breaking change exception.

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Generated/SdkInfo_BatchManagement.cs:
```diff
@@ -1,46 +0,0 @@
-
```
this file should still be generated, can you please include it?

paterasMSFT replied:
Why do we need it? It's internal and not referenced anywhere.

markcowl replied:
Honestly, not sure - It may be used by some tooling to discover api-versions, I do  know that it is required, and should be generated as part of running generation.

paterasMSFT replied:
Do you know if there's a flag I need to set when calling AutoRest or something to get it to generate? Not sure why it's not.

markcowl reviewed (changes requested):
There are some breaking changes here.  The changes aoround identity types I think can be easily mitigated by a customization, per the comments.

The breaking change in the constructor for ContainerRegisytry is caused by the way parameters are ordered (optional parameters come after required parameters).  I don't think there is an easy mitigation here that is desirable.

paterasMSFT committed:
Fixed breaking change

Fixed breacking change

paterasMSFT commented:
@markcowl I've addressed the breaking change issues (and removed the "breaking change" and "needs revision" labels). I think this is ready to go, pending your approval. If you're good with it, we'd love to get it merged as soon as possible.

Thanks!

paterasMSFT commented:
@markcowl could we get an approval here? Time is a factor. Thank you.

paterasMSFT committed:
Updated SDKInfo

Updated generation.

markcowl reviewed (changes requested):
It looks like you addressecd one breaking change issue, but did not address the other two.  Did I miss something?

paterasMSFT commented:
> still a breaking change - the constructor has removed parameters.

Which parameters have been removed? I still see them all there. The type name is shorter, will the old not be coerced into its parent type?

> Still a breaking change, this class is still removed with no mitigation

The generated one is removed, but I put in a manual one that derives from the newly renamed generated class, as discussed.

paterasMSFT committed:
Added BatchAccountIdentity constructor and test for equality against old type.

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Properties/AssemblyInfo.cs:
```diff
@@ -8,8 +8,8 @@
 
-[assembly: AssemblyVersion("13.0.0.0")]
-[assembly: AssemblyFileVersion("13.0.0.0")]
+[assembly: AssemblyVersion("15.0.0.0")]
```
This should match the package version

markcowl replied:
```suggestion
[assembly: AssemblyVersion("14.0.0.0")]
```

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Microsoft.Azure.Management.Batch.csproj:
```diff
@@ -8,7 +8,7 @@
     <AssemblyTitle>Microsoft Azure Batch Management Library</AssemblyTitle>
     <AssemblyName>Microsoft.Azure.Management.Batch</AssemblyName>
-    <Version>13.0.0</Version>
+    <Version>14.0.0</Version>
```
You don't need to update the major version, this could be 13.1.0, but this is up to you.  regardless, this version should match the assembly and assemblyFileVersion marked below.

paterasMSFT committed:
Fixed version.

Fixed version.

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Microsoft.Azure.Management.Batch.csproj:
```diff
@@ -8,7 +8,7 @@
     <AssemblyTitle>Microsoft Azure Batch Management Library</AssemblyTitle>
     <AssemblyName>Microsoft.Azure.Management.Batch</AssemblyName>
-    <Version>13.0.0</Version>
+    <Version>15.0.0</Version>
```
```suggestion
    <Version>14.0.0</Version>
```

markcowl commented on sdk/batch/Microsoft.Azure.Management.Batch/src/Properties/AssemblyInfo.cs:
```diff
@@ -8,8 +8,8 @@
-[assembly: AssemblyVersion("13.0.0.0")]
-[assembly: AssemblyFileVersion("13.0.0.0")]
+[assembly: AssemblyVersion("15.0.0.0")]
+[assembly: AssemblyFileVersion("15.0.0.0")]
```
```suggestion
[assembly: AssemblyFileVersion("14.0.0.0")]
```
